spatialos-sdk/examples/project-example/setup.rs

```diff
@@ -0,0 +1,81 @@
+use fs_extra::dir::{self, CopyOptions};
+use std::ffi::OsString;
+use std::fs;
+use std::path::PathBuf;
+use std::process::Command;
+use tap::*;
+
+pub fn do_setup(spatial_lib_dir: Option<PathBuf>, out_dir: PathBuf) {
+    let spatial_lib_dir = spatial_lib_dir
+        .or_else(|| std::env::var("SPATIAL_LIB_DIR").map(Into::into).ok())
+        .expect("--spatial_lib_dir argument must be specified or the SPATIAL_LIB_DIR environment variable must be set");
+
+    // Determine the paths the the schema compiler and protoc relative the the lib
+    // dir path.
+    let schema_compiler_path = spatial_lib_dir.join("schema-compiler/schema_compiler");
+    let protoc_path = spatial_lib_dir.join("schema-compiler/protoc");
+
+    // Calculate the various output directories relative to `out_dir`.
+    let bin_path = out_dir.join("spatialos/schema/bin");
+    let tmp_path = out_dir.join("tmp");
+
+    // Create the output directories if they don't already exist.
+    fs::create_dir_all(&bin_path).expect("Failed to crate spatialos/schema/bin");
+    fs::create_dir_all(&tmp_path).expect("Failed to create tmp");
+
+    // Copy the contents of the schema-compiler/proto dir into the temp dir.
+    let proto_dir_glob = spatial_lib_dir.join("schema-compiler/proto/*");
+    for entry in glob::glob(proto_dir_glob.to_str().unwrap())
+        .unwrap()
+        .filter_map(Result::ok)
+    {
+        dir::copy(
+            &entry,
+            &tmp_path,
+            &CopyOptions {
+                overwrite: true,
+                ..CopyOptions::new()
+            },
+        )
+        .expect("Failed to copy contents of schema-compiler/proto");
+    }
+
+    // Run the schema compiler for each of the schema files in std-lib/improbable.
+    let schema_glob = spatial_lib_dir.join("std-lib/improbable/*.schema");
+    let schema_path_arg =
+        OsString::from("--schema_path=").tap(|arg| arg.push(&spatial_lib_dir.join("std-lib")));
```

Comment:
TIL about `tap` - pretty neat 👍

Reply:
Yeah, it's super useful! If you want to read more about it, this is the blog post that talks about it more in detail: https://myrrlyn.net/blog/misc/rust-flow-part-two